File: sdk/go/sessions.go
package qeetid

import (
	"context"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math/big"
	"net/http"
	"strings"
	"sync"
	"time"
)

// Claims is the verified content of a Qeet-issued token.
type Claims struct {
	UserID    string
	TenantID  string
	SessionID string
	Scope     string
	Subject   string
	Issuer    string
	ExpiresAt int64
	IssuedAt  int64
	Raw       map[string]any
}

// VerifyOptions tightens verification. ClockSkew defaults to 30s.
type VerifyOptions struct {
	Issuer    string
	Audience  string
	ClockSkew time.Duration
}

// Sessions verifies ES256 tokens against the issuer's published JWKS. After the
// keys are cached it is fully local — the hosted-aligned way to identify the
// caller without a network round-trip per request.
type Sessions struct {
	jwksURL string
	hc      *http.Client

	mu        sync.Mutex
	keys      map[string]*ecdsa.PublicKey
	fetchedAt time.Time
}

const jwksTTL = 5 * time.Minute

func newSessions(baseURL string, hc *http.Client) *Sessions {
	return &Sessions{jwksURL: baseURL + "/.well-known/jwks.json", hc: hc}
}

type jwk struct {
	Kty string `json:"kty"`
	Crv string `json:"crv"`
	X   string `json:"x"`
	Y   string `json:"y"`
	Kid string `json:"kid"`
	Use string `json:"use"`
}

// Verify checks the token's ES256 signature against the JWKS, then validates
// expiry/issuer/audience. Returns *Claims or an error.
func (s *Sessions) Verify(ctx context.Context, token string, opts ...VerifyOptions) (*Claims, error) {
	var o VerifyOptions
	if len(opts) > 0 {
		o = opts[0]
	}
	skew := o.ClockSkew
	if skew == 0 {
		skew = 30 * time.Second
	}

	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, &Error{Status: 401, Code: "invalid_token", Message: "malformed token"}
	}
	var header struct {
		Alg string `json:"alg"`
		Kid string `json:"kid"`
	}
	if err := decodeSegment(parts[0], &header); err != nil {
		return nil, err
	}
	if header.Alg != "ES256" {
		return nil, &Error{Status: 401, Code: "invalid_token", Message: "unsupported alg " + header.Alg}
	}

	pub, err := s.resolveKey(ctx, header.Kid)
	if err != nil {
		return nil, err
	}

	sig, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil || len(sig) != 64 {
		return nil, &Error{Status: 401, Code: "invalid_token", Message: "malformed signature"}
	}
	r := new(big.Int).SetBytes(sig[:32])
	ss := new(big.Int).SetBytes(sig[32:])
	digest := sha256.Sum256([]byte(parts[0] + "." + parts[1]))
	if !ecdsa.Verify(pub, digest[:], r, ss) {
		return nil, &Error{Status: 401, Code: "invalid_token", Message: "signature verification failed"}
	}

	var raw map[string]any
	if err := decodeSegment(parts[1], &raw); err != nil {
		return nil, err
	}
	now := time.Now().Unix()
	sk := int64(skew / time.Second)
	exp := int64Claim(raw["exp"])
	if exp == 0 || now > exp+sk {
		return nil, &Error{Status: 401, Code: "invalid_token", Message: "token expired"}
	}
	if nbf := int64Claim(raw["nbf"]); nbf != 0 && now+sk < nbf {
		return nil, &Error{Status: 401, Code: "invalid_token", Message: "token not yet valid"}
	}
	if o.Issuer != "" && strClaim(raw["iss"]) != o.Issuer {
		return nil, &Error{Status: 401, Code: "invalid_token", Message: "issuer mismatch"}
	}
	if o.Audience != "" && !audienceMatches(raw["aud"], o.Audience) {
		return nil, &Error{Status: 401, Code: "invalid_token", Message: "audience mismatch"}
	}

	userID := strClaim(raw["user_id"])
	if userID == "" {
		userID = strClaim(raw["sub"])
	}
	return &Claims{
		UserID:    userID,
		TenantID:  strClaim(raw["tenant_id"]),
		SessionID: strClaim(raw["sid"]),
		Scope:     strClaim(raw["scope"]),
		Subject:   strClaim(raw["sub"]),
		Issuer:    strClaim(raw["iss"]),
		ExpiresAt: exp,
		IssuedAt:  int64Claim(raw["iat"]),
		Raw:       raw,
	}, nil
}

func (s *Sessions) resolveKey(ctx context.Context, kid string) (*ecdsa.PublicKey, error) {
	if k := s.lookup(kid, false); k != nil {
		return k, nil
	}
	if err := s.refresh(ctx); err != nil {
		return nil, err
	}
	if k := s.lookup(kid, true); k != nil {
		return k, nil
	}
	return nil, &Error{Status: 401, Code: "invalid_token", Message: "no JWKS key for kid " + kid}
}

func (s *Sessions) lookup(kid string, forceFresh bool) *ecdsa.PublicKey {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.keys == nil || (!forceFresh && time.Since(s.fetchedAt) > jwksTTL) {
		return nil
	}
	if kid == "" {
		for _, k := range s.keys {
			return k
		}
		return nil
	}
	return s.keys[kid]
}

func (s *Sessions) refresh(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.jwksURL, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	res, err := s.hc.Do(req)
	if err != nil {
		return &Error{Code: "network_error", Message: "JWKS fetch: " + err.Error()}
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		return &Error{Status: res.StatusCode, Code: "jwks_error", Message: "JWKS fetch failed"}
	}
	var doc struct {
		Keys []jwk `json:"keys"`
	}
	if err := json.NewDecoder(res.Body).Decode(&doc); err != nil {
		return &Error{Code: "jwks_error", Message: "JWKS decode: " + err.Error()}
	}
	keys := make(map[string]*ecdsa.PublicKey, len(doc.Keys))
	for _, k := range doc.Keys {
		pub, err := k.toECDSA()
		if err != nil {
			continue
		}
		keys[k.Kid] = pub
	}
	s.mu.Lock()
	s.keys = keys
	s.fetchedAt = time.Now()
	s.mu.Unlock()
	return nil
}

func (k jwk) toECDSA() (*ecdsa.PublicKey, error) {
	if k.Kty != "EC" || k.Crv != "P-256" {
		return nil, fmt.Errorf("unsupported key %s/%s", k.Kty, k.Crv)
	}
	xb, err := base64.RawURLEncoding.DecodeString(k.X)
	if err != nil {
		return nil, err
	}
	yb, err := base64.RawURLEncoding.DecodeString(k.Y)
	if err != nil {
		return nil, err
	}
	return &ecdsa.PublicKey{Curve: elliptic.P256(), X: new(big.Int).SetBytes(xb), Y: new(big.Int).SetBytes(yb)}, nil
}

func decodeSegment(seg string, out any) error {
	b, err := base64.RawURLEncoding.DecodeString(seg)
	if err != nil {
		return &Error{Status: 401, Code: "invalid_token", Message: "malformed token segment"}
	}
	if err := json.Unmarshal(b, out); err != nil {
		return &Error{Status: 401, Code: "invalid_token", Message: "malformed token segment"}
	}
	return nil
}

func strClaim(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func int64Claim(v any) int64 {
	if f, ok := v.(float64); ok {
		return int64(f)
	}
	return 0
}

func audienceMatches(aud any, want string) bool {
	switch a := aud.(type) {
	case string:
		return a == want
	case []any:
		for _, x := range a {
			if s, ok := x.(string); ok && s == want {
				return true
			}
		}
	}
	return false
}

File: sdk/python/qeetid/sessions.py
"""Session/JWT verification against the issuer's published JWKS.

Verifies Qeet-issued **ES256** (ECDSA P-256) tokens against the keys published
at ``<base_url>/.well-known/jwks.json``. After the keys are cached it is fully
local, so it's cheap to call on every request. Mirrors ``sdk/go/sessions.go``
(IEEE-P1363 raw-``r||s`` ECDSA verification) and the TS ``Sessions`` class.
"""

from __future__ import annotations

import base64
import json
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Union

import httpx
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec, utils

from .errors import SessionVerificationError

__all__ = ["Sessions", "SessionClaims", "VerifyOptions"]

JWKS_TTL_SECONDS = 5 * 60


@dataclass
class SessionClaims:
    """The verified content of a Qeet-issued token."""

    user_id: str
    subject: str
    expires_at: int
    tenant_id: Optional[str] = None
    session_id: Optional[str] = None
    scope: Optional[str] = None
    issuer: Optional[str] = None
    audience: Optional[Union[str, List[str]]] = None
    issued_at: Optional[int] = None
    raw: Dict[str, Any] = field(default_factory=dict)


@dataclass
class VerifyOptions:
    """Tightens verification. ``clock_tolerance_seconds`` defaults to 30."""

    issuer: Optional[str] = None
    audience: Optional[str] = None
    clock_tolerance_seconds: int = 30


class Sessions:
    """Verifies ES256 tokens against the issuer's published JWKS.

    The JWKS is fetched lazily on first use, cached for
    :data:`JWKS_TTL_SECONDS`, and refreshed once on an unknown ``kid`` (key
    rotation).
    """

    def __init__(self, base_url: str, http_client: httpx.Client) -> None:
        self._jwks_url = base_url.rstrip("/") + "/.well-known/jwks.json"
        self._http = http_client
        self._keys: Optional[Dict[str, ec.EllipticCurvePublicKey]] = None
        self._fetched_at = 0.0

    def verify(
        self, token: str, options: Optional[VerifyOptions] = None
    ) -> SessionClaims:
        """Verify ``token``'s ES256 signature against the JWKS, then validate
        expiry / not-before / issuer / audience. Returns :class:`SessionClaims`
        or raises :class:`SessionVerificationError`.
        """
        opts = options or VerifyOptions()
        skew = opts.clock_tolerance_seconds if opts.clock_tolerance_seconds else 30

        parts = token.split(".")
        if len(parts) != 3:
            raise SessionVerificationError("malformed token")
        h, p, s = parts

        header = _decode_segment(h)
        if header.get("alg") != "ES256":
            raise SessionVerificationError(f"unsupported alg {header.get('alg')}")
        kid = header.get("kid") if isinstance(header.get("kid"), str) else None

        key = self._resolve_key(kid)

        sig = _b64url_decode(s)
        if len(sig) != 64:
            raise SessionVerificationError("malformed signature")
        # JWS ES256 carries the raw IEEE-P1363 r||s signature; convert it to the
        # DER form the cryptography library verifies, exactly like the Go SDK
        # splits sig[:32]/sig[32:] into big.Ints.
        r = int.from_bytes(sig[:32], "big")
        ss = int.from_bytes(sig[32:], "big")
        der = utils.encode_dss_signature(r, ss)
        try:
            key.verify(der, f"{h}.{p}".encode("ascii"), ec.ECDSA(hashes.SHA256()))
        except InvalidSignature as exc:
            raise SessionVerificationError("signature verification failed") from exc

        payload = _decode_segment(p)
        now = int(time.time())

        exp = _int_claim(payload.get("exp"))
        if exp is None or now > exp + skew:
            raise SessionVerificationError("token expired")
        nbf = _int_claim(payload.get("nbf"))
        if nbf is not None and now + skew < nbf:
            raise SessionVerificationError("token not yet valid")
        if opts.issuer and payload.get("iss") != opts.issuer:
            raise SessionVerificationError("issuer mismatch")
        if opts.audience and not _audience_matches(payload.get("aud"), opts.audience):
            raise SessionVerificationError("audience mismatch")

        user_id = _str_claim(payload.get("user_id")) or _str_claim(payload.get("sub")) or ""
        return SessionClaims(
            user_id=user_id,
            tenant_id=_str_claim(payload.get("tenant_id")),
            session_id=_str_claim(payload.get("sid")),
            scope=_str_claim(payload.get("scope")),
            subject=_str_claim(payload.get("sub")) or "",
            issuer=_str_claim(payload.get("iss")),
            audience=payload.get("aud"),
            expires_at=exp,
            issued_at=_int_claim(payload.get("iat")),
            raw=payload,
        )

    # ---- key resolution / JWKS ---------------------------------------------
    def _resolve_key(self, kid: Optional[str]) -> ec.EllipticCurvePublicKey:
        key = self._lookup(kid, force_fresh=False)
        if key is None:
            self._refresh()
            key = self._lookup(kid, force_fresh=True)
        if key is None:
            raise SessionVerificationError(
                f"no JWKS key for kid {kid}" if kid else "no usable JWKS key"
            )
        return key

    def _lookup(
        self, kid: Optional[str], force_fresh: bool
    ) -> Optional[ec.EllipticCurvePublicKey]:
        if self._keys is None or (
            not force_fresh and time.time() - self._fetched_at > JWKS_TTL_SECONDS
        ):
            return None
        if not kid:
            return next(iter(self._keys.values()), None)
        return self._keys.get(kid)

    def _refresh(self) -> None:
        try:
            res = self._http.get(self._jwks_url, headers={"Accept": "application/json"})
        except httpx.HTTPError as exc:
            raise SessionVerificationError(f"JWKS fetch failed: {exc}") from exc
        if res.status_code != 200:
            raise SessionVerificationError(f"JWKS fetch failed: {res.status_code}")
        try:
            doc = res.json()
        except ValueError as exc:
            raise SessionVerificationError(f"JWKS decode failed: {exc}") from exc

        keys: Dict[str, ec.EllipticCurvePublicKey] = {}
        for jwk in doc.get("keys", []) or []:
            pub = _jwk_to_ec(jwk)
            if pub is not None:
                keys[jwk.get("kid", "")] = pub
        self._keys = keys
        self._fetched_at = time.time()


def _jwk_to_ec(jwk: Dict[str, Any]) -> Optional[ec.EllipticCurvePublicKey]:
    if jwk.get("kty") != "EC" or jwk.get("crv") != "P-256":
        return None
    try:
        x = int.from_bytes(_b64url_decode(jwk["x"]), "big")
        y = int.from_bytes(_b64url_decode(jwk["y"]), "big")
    except (KeyError, ValueError):
        return None
    try:
        return ec.EllipticCurvePublicNumbers(x, y, ec.SECP256R1()).public_key()
    except ValueError:
        return None


def _decode_segment(seg: str) -> Dict[str, Any]:
    try:
        out = json.loads(_b64url_decode(seg))
    except ValueError as exc:
        raise SessionVerificationError("malformed token segment") from exc
    if not isinstance(out, dict):
        raise SessionVerificationError("malformed token segment")
    return out


def _b64url_decode(seg: str) -> bytes:
    padding = "=" * (-len(seg) % 4)
    return base64.urlsafe_b64decode(seg + padding)


def _str_claim(v: Any) -> Optional[str]:
    return v if isinstance(v, str) else None


def _int_claim(v: Any) -> Optional[int]:
    # JSON numbers decode to int/float; bool is a subclass of int, exclude it.
    if isinstance(v, bool):
        return None
    if isinstance(v, (int, float)):
        return int(v)
    return None


def _audience_matches(aud: Any, want: str) -> bool:
    if isinstance(aud, str):
        return aud == want
    if isinstance(aud, list):
        return want in aud
    return False
